From 4ea7be222374301c50f6ededcc4addcda78c1c28 Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Wed, 15 Apr 2020 00:02:08 +0300
Subject: [PATCH] [SW-0] hl: add clear of bss to crt0

Now all global not initialized variables can be assumed to have value of
zero upon first read access

Change-Id: Iddac4f8eca1067c73409595e04bc9609e35e50d0
---

diff --git a/app/crt0_preboot.S b/app/crt0_preboot.S
index 5bb2238..f2608b1 100644
--- a/app/crt0_preboot.S
+++ b/app/crt0_preboot.S
@@ -19,12 +19,28 @@
 
 	mov	sp, __stack_top
 	mov	fp, sp
+
+/* Clear BSS */
+	mov	r5, __bss_start
+	sub	r6, __bss_end, r5
+	lsr.f	lp_count, r6, 2
+	lpnz	1f
+	st.ab   0, [r5, 4]
+1:
 #endif
 
 #ifdef __aarch64__
 	ldr	x0, =__stack_top
 	bic	sp, x0, #0xf	/* 16-byte alignment for ABI compliance */
 	mov	x29, sp		/* Set FP (x29) */
+
+/* Clear BSS */
+	ldr	x0, =__bss_start
+	ldr	x1, =__bss_end
+clear_loop:
+	str	xzr, [x0], #8
+	cmp	x0, x1
+	b.lo	clear_loop
 #endif
 
 	bl	preboot_main	/* Set LR (x30) */
diff --git a/gaudi/preboot.lds.S b/gaudi/preboot.lds.S
index 01650fd..b244777 100644
--- a/gaudi/preboot.lds.S
+++ b/gaudi/preboot.lds.S
@@ -20,6 +20,13 @@
 		* (.text*)
 	} >.sram
 
+	.bss : {
+		. = ALIGN(8);
+		__bss_start = . ;
+		*(.bss*)
+		__bss_end = . ;
+	} > .sram
+
 	.stack : {
 		. = ALIGN(8);
 		__stack = . ;
diff --git a/goya/preboot.lds.S b/goya/preboot.lds.S
index 01650fd..b244777 100644
--- a/goya/preboot.lds.S
+++ b/goya/preboot.lds.S
@@ -20,6 +20,13 @@
 		* (.text*)
 	} >.sram
 
+	.bss : {
+		. = ALIGN(8);
+		__bss_start = . ;
+		*(.bss*)
+		__bss_end = . ;
+	} > .sram
+
 	.stack : {
 		. = ALIGN(8);
 		__stack = . ;
diff --git a/goya2/preboot.lds.S b/goya2/preboot.lds.S
index 01650fd..b244777 100644
--- a/goya2/preboot.lds.S
+++ b/goya2/preboot.lds.S
@@ -20,6 +20,13 @@
 		* (.text*)
 	} >.sram
 
+	.bss : {
+		. = ALIGN(8);
+		__bss_start = . ;
+		*(.bss*)
+		__bss_end = . ;
+	} > .sram
+
 	.stack : {
 		. = ALIGN(8);
 		__stack = . ;
diff --git a/goya2_arc/preboot.lds.S b/goya2_arc/preboot.lds.S
index 03c9bf9..2950588 100644
--- a/goya2_arc/preboot.lds.S
+++ b/goya2_arc/preboot.lds.S
@@ -22,6 +22,13 @@
 		* (.text*)
 	} >.sram
 
+	.bss : {
+		. = ALIGN(8);
+		__bss_start = . ;
+		*(.bss*)
+		__bss_end = . ;
+	} > .sram
+
 	.stack : {
 		. = ALIGN(8);
 		__stack = . ;
